Deploy fails on "node-pre-gyp install --fallback-to-build"

Hi there,

deploy fails with following error message (issue seems to be htpasswd-auth):

```
npm WARN package.json lambda-cloudfront-cookies@1.0.0 No repository field.
npm WARN package.json lambda-cloudfront-cookies@1.0.0 No README data
WARN engine apache-crypt@1.2.1: wanted: {"node":">=4.6.1"} (current: {"node":"4.3.2","npm":"2.14.12"})
WARN engine apache-md5@1.1.2: wanted: {"node":">=4.6.1"} (current: {"node":"4.3.2","npm":"2.14.12"})

> bcrypt@2.0.1 install /build/node_modules/htpasswd-auth/node_modules/bcrypt
> node-pre-gyp install --fallback-to-build

node-pre-gyp ERR! UNCAUGHT EXCEPTION 
node-pre-gyp ERR! stack TypeError: this is not a typed array.
node-pre-gyp ERR! stack     at Function.from (native)
node-pre-gyp ERR! stack     at Object.<anonymous> (/build/node_modules/htpasswd-auth/node_modules/bcrypt/node_modules/tar/lib/parse.js:33:27)
node-pre-gyp ERR! stack     at Module._compile (module.js:409:26)
node-pre-gyp ERR! stack     at Object.Module._extensions..js (module.js:416:10)
node-pre-gyp ERR! stack     at Module.load (module.js:343:32)
node-pre-gyp ERR! stack     at Function.Module._load (module.js:300:12)
node-pre-gyp ERR! stack     at Module.require (module.js:353:17)
node-pre-gyp ERR! stack     at require (internal/module.js:12:17)
node-pre-gyp ERR! stack     at Object.<anonymous> (/build/node_modules/htpasswd-auth/node_modules/bcrypt/node_modules/tar/lib/list.js:10:16)
node-pre-gyp ERR! stack     at Module._compile (module.js:409:26)
node-pre-gyp ERR! System Linux 4.9.87-linuxkit-aufs
node-pre-gyp ERR! command "/usr/local/lib64/node-v4.3.x/bin/node" "/build/node_modules/htpasswd-auth/node_modules/bcrypt/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build"
node-pre-gyp ERR! cwd /build/node_modules/htpasswd-auth/node_modules/bcrypt
node-pre-gyp ERR! node -v v4.3.2
node-pre-gyp ERR! node-pre-gyp -v v0.9.1
node-pre-gyp ERR! This is a bug in `node-pre-gyp`.
node-pre-gyp ERR! Try to update node-pre-gyp and file an issue if it does not help:
node-pre-gyp ERR!     <https://github.com/mapbox/node-pre-gyp/issues>
npm ERR! Linux 4.9.87-linuxkit-aufs
npm ERR! argv "/usr/local/lib64/node-v4.3.x/bin/node" "/usr/local/lib64/node-v4.3.x/bin/npm" "install" "--save" "htpasswd-auth"
npm ERR! node v4.3.2
npm ERR! npm  v2.14.12
npm ERR! code ELIFECYCLE

npm ERR! bcrypt@2.0.1 install: `node-pre-gyp install --fallback-to-build`
npm ERR! Exit status 7
npm ERR! 
npm ERR! Failed at the bcrypt@2.0.1 install script 'node-pre-gyp install --fallback-to-build'.
npm ERR! This is most likely a problem with the bcrypt package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-pre-gyp install --fallback-to-build
npm ERR! You can get their info via:
npm ERR!     npm owner ls bcrypt
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /build/npm-debug.log
The command '/bin/sh -c npm install --save htpasswd-auth' returned a non-zero code: 1
```

So i changed the dockerfile to:
```
FROM lambci/lambda-base:build

ENV PATH=/var/lang/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin \
    LD_LIBRARY_PATH=/var/lang/lib:/lib64:/usr/lib64:/var/runtime:/var/runtime/lib:/var/task:/var/task/lib \
    AWS_EXECUTION_ENV=AWS_Lambda_nodejs8.10 \
    NODE_PATH=/var/runtime:/var/task:/var/runtime/node_modules \
    npm_config_unsafe-perm=true

RUN rm -rf /var/runtime /var/lang && \
  curl https://lambci.s3.amazonaws.com/fs/nodejs8.10.tgz | tar -zx -C /

CMD ["npm", "rebuild"]

RUN mkdir /build
WORKDIR /build
COPY package.json /build/package.json

RUN npm install --production

COPY index.js /build/index.js

RUN zip -r /build/dist.zip . > /dev/null
CMD ["cat", "/build/dist.zip"]
```

Which results in a finished deploy. Taking a look at the stack everything seems fine, but I'm unable to login with my credentials on the site.

Hi all, can you link me to the discussion you all started with bcrypt after following the instructions in the error log?

```
npm ERR! Failed at the bcrypt@2.0.1 install script 'node-pre-gyp install --fallback-to-build'.
npm ERR! This is most likely a problem with the bcrypt package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-pre-gyp install --fallback-to-build
npm ERR! You can get their info via:
npm ERR!     npm owner ls bcrypt
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /build/npm-debug.log
The command '/bin/sh -c npm install --save htpasswd-auth' returned a non-zero code: 1
```

--- a/.gitignore
+++ b/.gitignore
@@ -2,3 +2,4 @@ node_modules/
 dist/
 .sass-cache/
 *.css.map
+.idea/

--- a/README.md
+++ b/README.md
@@ -131,6 +131,13 @@ It should contain the following info - minus the comments:
 
   // encrypted contents of the <htpasswd> file from step 6
   "encryptedHtpasswd=AQICAH..."
+  
+  // ------------------
+  // SSL Certificate ARN
+  // - provide this if you want to use an existing ACM Certificate.
+  // - see below in the README
+  // ------------------
+  "sslCertificateArn=arn:aws:acm:us-east-1..."
 ]
 ```
 
@@ -139,7 +146,7 @@ You can then deploy the full stack using:
 ```bash
 # name of an S3 bucket for storing the Lambda code
 # bucket will be created if it doesn't already exist
-./deploy awspics-lambda
+./deploy <unique_bucket_name_here>
 ```
 
 Any subsequent changes that you make to this code can be redeployed with the
@@ -149,6 +156,13 @@ modified resources.
 The first deployment should take about 30 minutes since there's a lot to set up.
 You'll also receive an email to approve the SSL certificate request, which you
 should complete quickly, so that the rest of the deployment can proceed.
+
+##### Note on SSL Cert
+AWS Certificate Manager now supports SSL cert verification via DNS validation.
+It is recommended that you manually request the certificate for your hosted zone and
+ chose DNS validation method for much faster validation. Then use the resulting ARN 
+ in your config. You can also leave this config key empty to create the certificate as 
+ normal.
 
 Once the initial deployment is done, you'll need to point your domain's DNS
 settings to add a CNAME to the newly created CloudFront Distribution URL, which

--- a/app.yml
+++ b/app.yml
@@ -52,7 +52,7 @@ Resources:
     Type: AWS::Serverless::Function
     Properties:
       Handler: index.handler
-      Runtime: nodejs4.3
+      Runtime: nodejs8.10
       CodeUri: dist/lambda-login.zip
       Role: !GetAtt LambdaRole.Arn
       Environment:

--- a/app.yml
+++ b/app.yml
@@ -42,6 +42,12 @@ Parameters:
     Description: htpasswd file contents encrypted with KMS
     Type: String
     NoEcho: true
+  sslCertificateArn:
+    Description: ACM Certificate ARN
+    Type: String
+
+Conditions:
+  createSSLCert: !Equals [ !Ref sslCertificateArn, "" ]
 
 Resources:
 
@@ -328,6 +334,7 @@ Resources:
   #
   SSLCert:
     Type: AWS::CertificateManager::Certificate
+    Condition: createSSLCert
     Properties:
       DomainName: !Ref website
 
@@ -422,5 +429,5 @@ Resources:
           ViewerProtocolPolicy: redirect-to-https
         PriceClass: PriceClass_All
         ViewerCertificate:
-          AcmCertificateArn: !Ref SSLCert
+          AcmCertificateArn: !If [ createSSLCert, !Ref SSLCert, !Ref sslCertificateArn ]
           SslSupportMethod: sni-only

--- a/config.example.json
+++ b/config.example.json
@@ -9,5 +9,6 @@
   "kmsKeyId=00000000-0000-0000-0000-000000000000",
   "cloudFrontKeypairId=APK...",
   "encryptedCloudFrontPrivateKey=AQICAH...",
-  "encryptedHtpasswd=AQICAH..."
+  "encryptedHtpasswd=AQICAH...",
+  "sslCertificateArn="
 ]

--- a/login/Dockerfile
+++ b/login/Dockerfile
@@ -1,4 +1,4 @@
-FROM lambci/lambda:build
+FROM lambci/lambda:build:build-nodejs8.10
 
 # working folder
 RUN mkdir /build

--- a/login/Dockerfile
+++ b/login/Dockerfile
@@ -1,4 +1,4 @@
-FROM lambci/lambda:build:build-nodejs8.10
+FROM lambci/lambda:build-nodejs8.10
 
 # working folder
 RUN mkdir /build

--- a/login/package.json
+++ b/login/package.json
@@ -8,6 +8,8 @@
   "dependencies": {
     "async": "^2.1.4",
     "aws-cloudfront-sign": "^2.2.0",
-    "htpasswd-auth": "^2.0.0"
+    "content-type": "^1.0.2",
+    "htpasswd-auth": "^2.0.0",
+    "querystringparser": "^0.1.1"
   }
 }

--- a/resize/Dockerfile
+++ b/resize/Dockerfile
@@ -1,4 +1,4 @@
-FROM lambci/lambda:build
+FROM lambci/lambda:build:build-nodejs8.10
 
 # working folder
 RUN mkdir /build

--- a/resize/Dockerfile
+++ b/resize/Dockerfile
@@ -1,4 +1,4 @@
-FROM lambci/lambda:build:build-nodejs8.10
+FROM lambci/lambda:build-nodejs8.10
 
 # working folder
 RUN mkdir /build

--- a/site-builder/Dockerfile
+++ b/site-builder/Dockerfile
@@ -1,4 +1,4 @@
-FROM lambci/lambda:build
+FROM lambci/lambda:build-nodejs8.10
 
 # working folder
 RUN mkdir /build

